File: 0020-valid-parentheses/solution.java
/*
 * LeetCode #20 - Valid Parentheses
 * Difficulty : Easy
 * Language   : java
 * Runtime    : 3 ms
 * Memory     : 43.13 MB
 * URL        : https://leetcode.com/problems/valid-parentheses/
 */

import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (char ch : s.toCharArray()) {
            if (ch == '(' || ch == '{' || ch == '[') {
                stack.push(ch);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }

                char top = stack.pop();

                if (ch == ')' && top != '(' ||
                    ch == '}' && top != '{' ||
                    ch == ']' && top != '[') {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }
}

File: 0020-valid-parentheses/solution.py
# LeetCode #20 - Valid Parentheses
# Difficulty : Easy
# Language   : python
# Runtime    : 0 ms
# Memory     : 19.09 MB
# URL        : https://leetcode.com/problems/valid-parentheses/

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        pairs = {
            ')': '(',
            '}': '{',
            ']': '['
        }

        for ch in s:
            if ch in "({[":
                stack.append(ch)
            else:
                if not stack or stack[-1] != pairs[ch]:
                    return False
                stack.pop()

        return len(stack) == 0